modeshape-jcr/src/main/java/org/modeshape/jcr/spi/index/provider/IndexChangeAdapters.java

```diff
@@ -334,26 +439,35 @@ protected void removeNode( String workspaceName,
                                    Path path,
                                    Name primaryType,
                                    Set<Name> mixinTypes ) {
-            if (includeRoot || path.isRoot()) {
-                index.remove(nodeKey(key));
-            }
+            index().remove(nodeKey(key));
         }
 
         @Override
-        protected void completeChanges() {
-            index.commit();
-            super.completeChanges();
-        }
-
-        @Override
-        protected void completeWorkspaceChanges() {
-            index.commit();
-            super.completeWorkspaceChanges();
+        protected void renameNode( String workspaceName, 
+                                   NodeKey key, 
+                                   Path newPath, 
+                                   Path.Segment oldSegment, 
+                                   Name primaryType,
+                                   Set<Name> mixinTypes ) {
+            PathFactory pathFactory = context.getValueFactories().getPathFactory();
+            Path oldPath = pathFactory.create(newPath.subpath(0, newPath.size()), oldSegment);
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
         }
 
         @Override
-        public String toString() {
-            return getClass().getSimpleName() + "(\"" + index.getName() + "\")";
+        protected void reorderNode( String workspaceName, 
+                                    NodeKey key, 
+                                    Name primaryType, 
+                                    Set<Name> mixinTypes, 
+                                    NodeKey parent,
+                                    Path newPath, 
+                                    Path oldPath, 
+                                    Path reorderedBeforePath ) {
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
```

Comment:
well this sucks because I don't see how this case can be handled: each reordering event contains only information about the node which was reordered - i.e. which node, what was the old path and what is the new path - which is what the index attempts to handle.

however, if this causes the side affect of altering the node paths of other SNS siblings which are also present in the index, there's is no way to make any changes for those as there's simply not enough information to do so.

So unless I'm missing something, I don't see any other options than raise a JIRA and leave it open :) (I'm also pretty sure the same problem exists in ModeShape 3).